sklearn/preprocessing/data.py

```diff
@@ -2555,3 +2559,201 @@ def quantile_transform(X, axis=0, n_quantiles=1000,
+      using the unique values.
+    sklearn.feature_extraction.DictVectorizer : performs a one-hot encoding of
+      dictionary items (also handles string-valued features).
+    sklearn.feature_extraction.FeatureHasher : performs an approximate one-hot
+      encoding of dictionary items or strings.
+    """
+
+    def __init__(self, encoding='onehot', categories='auto', dtype=np.float64,
+                 handle_unknown='error'):
+        self.encoding = encoding
+        self.categories = categories
+        self.dtype = dtype
+        self.handle_unknown = handle_unknown
+
+    def fit(self, X, y=None):
+        """Fit the CategoricalEncoder to X.
+
+        Parameters
+        ----------
+        X : array-like, shape [n_samples, n_feature]
+            The data to determine the categories of each feature.
+
+        Returns
+        -------
+        self
+        """
+
+        if self.encoding not in ['onehot', 'onehot-dense', 'ordinal']:
+            template = ("encoding should be either 'onehot', 'onehot-dense' "
+                        "or 'ordinal', got %s")
+            raise ValueError(template % self.handle_unknown)
+
+        if self.handle_unknown not in ['error', 'ignore']:
+            template = ("handle_unknown should be either 'error' or "
+                        "'ignore', got %s")
+            raise ValueError(template % self.handle_unknown)
+
+        if self.encoding == 'ordinal' and self.handle_unknown == 'ignore':
+            raise ValueError("handle_unknown='ignore' is not supported for"
+                             " encoding='ordinal'")
+
+        X = check_array(X, dtype=np.object, accept_sparse='csc', copy=True)
+        n_samples, n_features = X.shape
+
+        self._label_encoders_ = [LabelEncoder() for _ in range(n_features)]
+
+        for i in range(n_features):
+            le = self._label_encoders_[i]
+            Xi = X[:, i]
+            if self.categories == 'auto':
+                le.fit(Xi)
+            else:
+                valid_mask = np.in1d(Xi, self.categories[i])
+                if not np.all(valid_mask):
+                    if self.handle_unknown == 'error':
+                        diff = np.unique(Xi[~valid_mask])
+                        msg = ("Found unknown categories {0} in column {1}"
+                               " during fit".format(diff, i))
+                        raise ValueError(msg)
+                le.classes_ = np.array(np.sort(self.categories[i]))
+
+        self.categories_ = [le.classes_ for le in self._label_encoders_]
+
+        return self
+
+    def transform(self, X):
+        """Transform X using one-hot encoding.
+
+        Parameters
+        ----------
+        X : array-like, shape [n_samples, n_features]
+            The data to encode.
+
+        Returns
+        -------
+        X_out : sparse matrix or a 2-d array
+            Transformed input.
+
+        """
+        X = check_array(X, accept_sparse='csc', dtype=np.object, copy=True)
```

Comment:
Is this tested with sparse matrices? huh...

Reply:
why is ``dtype=np.object``? Shouldn't it be something like ``[np.object, np.int, somethingstring]``?

Reply:
> Is this tested with sparse matrices? huh...

Hmm, this indeed does not seem to be tested (only the sparse output, which is something different). This is something from the PR I took over. 
This might also be something we can leave for a follow-up? As this does not yet work out of the box (the LabelEncoder used under the hood does not work with sparse data. In principle I am planning to replace the use of this, but that was what we would postpone for a follow-up PR) I could convert sparse to dense before factorizing it, but that does defeat the purpose of using sparse a bit ..

> why is dtype=np.object? Shouldn't it be something like [np.object, np.int, somethingstring]?

Again, this is from the PR I took over, and didn't really thought about it. But actually, can't we just preserve the data type? (so `dtype=None`) ?

Reply:
> But actually, can't we just preserve the data type? (so dtype=None) ?

Ah, no, that does not work if you pass a list with mixed types (typically used in small examples). So something like `[np.object, np.int, somethingstring]` is probably better. The question to me is however what to include in this list? As actually almost any dtype is allowed (as long as you can take the unique values of it, and it is sortable). So in principle also a feature with some unique float values is OK. It is just if there is *no* dtype (eg list) that I want object.

Reply:
Related question: I am not sure the `copy=True` is needed? I don't think there is a risk on manipulating the values somewhere while encoding them.

Reply:
@amueller Can you look at this one?

So basically I want something like "preserve dtype, but if no dtype, then convert to object", which does not seems easy to express with `check_array` (although this is only to support lists, so not sure how important that is apart from examples)

Reply:
don't provide a dtype and if it comes out as string convert it to object?

Reply:
As I said, it's only for supporting lists (and actually even mixed type lists, so even more special case, although typically used in small examples), so not sure if it is that important. 
But the problem, in such case, with not specifying the dtype, is that a list of mixed types everything gets coerced to string (so also the integers).

Reply:
Right, I forgot that one... I remember discussing this at some length at several points. How about we convert without dtype, if the resulting dtype is string, we convert again with dtype object?
If someone gives us lists of stings we convert twice, but really, I don't think we can do better.

Reply:
@amueller The last commit tries this (convert, and if resulting dtype is string (and original dtype was not string), convert again with object dtype): https://github.com/scikit-learn/scikit-learn/pull/9151/commits/b40bd8e4eb96fa9514fdb955af2fe1b40661b1fe

If this is ok, maybe I should put it in a class method to avoid code duplication